src/System.Memory/tests/MemoryMarshal/CreateFromPinnedArray.cs

```diff
@@ -0,0 +1,160 @@
+        {
+            int[] a = new int[3];
+            Assert.Throws<ArgumentOutOfRangeException>(() => MemoryMarshal.CreateFromPinnedArray(a, 4, 0));
+        }
+
+        [Fact]
+        public static void CreateFromPinnedArrayWithStartAndNegativeLength()
+        {
+            int[] a = new int[3];
+            Assert.Throws<ArgumentOutOfRangeException>(() => MemoryMarshal.CreateFromPinnedArray(a, 0, -1));
+        }
+
+        [Fact]
+        public static void CreateFromPinnedArrayWithStartAndLengthTooLarge()
+        {
+            int[] a = new int[3];
+            Assert.Throws<ArgumentOutOfRangeException>(() => MemoryMarshal.CreateFromPinnedArray(a, 3, 1));
+            Assert.Throws<ArgumentOutOfRangeException>(() => MemoryMarshal.CreateFromPinnedArray(a, 2, 2));
+            Assert.Throws<ArgumentOutOfRangeException>(() => MemoryMarshal.CreateFromPinnedArray(a, 1, 3));
+            Assert.Throws<ArgumentOutOfRangeException>(() => MemoryMarshal.CreateFromPinnedArray(a, 0, 4));
+            Assert.Throws<ArgumentOutOfRangeException>(() => MemoryMarshal.CreateFromPinnedArray(a, int.MaxValue, int.MaxValue));
+        }
+
+        [Fact]
+        public static void CreateFromPinnedArrayWithStartAndLengthBothEqual()
+        {
+            // Valid for start to equal the array length. This returns an empty memory that starts "just past the array."
+            int[] a = { 91, 92, 93 };
+            Memory<int> pinnedMemory = MemoryMarshal.CreateFromPinnedArray(a, 3, 0);
+            pinnedMemory.Validate();
+        }
+
+        [Fact]
+        public static unsafe void CreateFromPinnedArrayVerifyPinning()
+        {
+            int[] unpinnedArray = { 90, 91, 92, 93, 94, 95, 96, 97, 98 };
+            int[] pinnedArray = { 90, 91, 92, 93, 94, 95, 96, 97, 98 };
+            GCHandle handle = GCHandle.Alloc(pinnedArray, GCHandleType.Pinned);
+
+            var memory = new Memory<int>(unpinnedArray, 0, 2);
+            void* ptr = Unsafe.AsPointer(ref MemoryMarshal.GetReference(memory.Span));
+
+            MemoryHandle memoryHandle = memory.Pin();
+            void* memoryHandlePtr = memoryHandle.Pointer;
+            memoryHandle.Dispose();
+
+            Memory<int> pinnedMemory = MemoryMarshal.CreateFromPinnedArray(pinnedArray, 0, 2);
+            void* pinnedPtr = Unsafe.AsPointer(ref MemoryMarshal.GetReference(pinnedMemory.Span));
+            void* memoryHandlePinnedPtr = pinnedMemory.Pin().Pointer;
+
+            GC.Collect();
+            GC.Collect(2);
+
+            Assert.Equal((int)pinnedPtr, (int)Unsafe.AsPointer(ref MemoryMarshal.GetReference(pinnedMemory.Span)));
+            Assert.NotEqual((int)ptr, (int)Unsafe.AsPointer(ref MemoryMarshal.GetReference(memory.Span)));
+
+            Assert.Equal((int)memoryHandlePinnedPtr, (int)handle.AddrOfPinnedObject().ToPointer());
+            Assert.NotEqual((int)memoryHandlePtr, (int)memory.Pin().Pointer);
+        }
+    }
+}
```

Comment:
It would be good to have a test that verifies that when memory is created from a prepinned array, it does not cause additional GCHandles to be created during Memory.Pin calls. I am not sure how to write such test, but maybe somebody on the runtime team would know. @jkotas?

Reply:
E.g. you would need to start the test in separate process, enable tracing for it, and verify via trace that there was no GC handle allocated. I do not think it is worth the effort with what we have available today. Writing similar tests may become more viable once we have the ability for processes to self-trace.

Reply:
> Writing similar tests may become more viable once we have the ability for processes to self-trace.

FWIW, we do something similar in a variety of corefx test suites for EventSource-based tracing, e.g.
https://github.com/dotnet/corefx/blob/master/src/System.Net.Security/tests/FunctionalTests/LoggingTest.cs#L31
but assuming GC handle-related tracing doesn't come from an EventSource, as Jan suggests that's going to be a lot more involved.